### DISCUSSION TITLE
any plan to make android side?

### DISCUSSION THREAD

**Discussion by mokhtarabadi** [eycorsican/leaf#30] [Q&A]
hi, did you have any plan to create the android side based on android VPN service?

**Comment by mokhtarabadi** [comment]
the leaf is a very awesome tool, I used it for more than 1 month on my Linux desktop, and also I used it on android by using termux, but for now I want to develop an android VPN based on the leaf, I used shadowsocks android source code, but because I do not know rust language, I can't write an RPC in rust and android to protecting leaf connection from android VPN service 
I need something like this
https://github.com/shadowsocks/shadowsocks-rust/blob/be6049f35bf32fa6e1e7df3e03a5b0bd9946fe9a/crates/shadowsocks/src/relay/sys/unix/mod.rs#L39

**Reply by mokhtarabadi** [reply]
@eycorsican 
it's great if you do it, I'll write the android client and publish the source code for anyone who needs leaf on their android device

**Reply by mokhtarabadi** [reply]
also for badvpn tun2socks, I think if inject the codes needed to tun2socks works on android (mean android VPN service give us fd instead of raw access to the tuntap device), I don't need the shadowsocks android source code (for tun2socks and send VPN service fd to it)

**Reply by eycorsican** [reply]
I think we're almost there. You got a [global dial function for TCP](https://github.com/eycorsican/leaf/blob/928ec6f0ebabe9b60e3420ab8ef9ca9661c8ba62/leaf/src/proxy/mod.rs#L100) and you just need to copy some code from shadowsocks-rust, that's almost done for the Rust part. I don't want to touch Android dev at the moment, I'm afraid you won't see my implementation in the near future.

**Reply by mokhtarabadi** [reply]
@eycorsican, thanks for your response

I'll try to implement the rust part...

one thing I needed just the custom HTTP headers in WebSocket (if you have free time please coding it!) and one more question, can I build leaf without tun2socks support? (mean remove the rust-tun code)

**Reply by eycorsican** [reply]
All inbounds and outbounds are optional modules, you could choose which modules to be included at compile time, please checkout "Customizing Build" section in readme.

I will have a look at header customization for ws.
